src/shims/fs.rs

```diff
@@ -193,21 +217,26 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
-
-        this.remove_handle_and(fd, |mut handle, this| {
-            // FIXME: Don't use raw methods
-            let bytes = this.memory.get_raw(buf.alloc_id).and_then(|alloc| {
-                alloc
-                    .get_bytes(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|bytes| handle.file.write(bytes).map(|bytes| bytes as i64))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?)
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            let bytes = this.memory.read_bytes(buf, Size::from_bytes(count))?;
+            let result = handle.file.write(&bytes);
+
+            match result {
+                Ok(c) => i64::try_from(c)
+                    .map_err(|_| err_unsup_format!("Number of written bytes {} cannot be transformed to i64", c).into()),
```

Comment:
Again when can this happen? I think like for `read` we should cap `count` to `1u64 << (pointer_size.bits() - 1)`, then this here cannot happen.